### ISSUE TITLE
The save_impressions background job is excruciatingly slow

### ISSUE THREAD

**Issue by jace** [hasgeek/hasjob#247]
`save_impressions` ([current version here](https://github.com/hasgeek/hasjob/blob/142400617d1f57306b53f700563f285117482e4d/hasjob/views/helper.py#L463-L498)) currently takes nearly a second per impression saved, so a typical home page view with ~30 jobs can tie up the background job runner for half a minute.

We don't know which of several factors is responsible for it being so slow, so this ticket is to help identify and fix the cause.

1. A simple query like `COUNT (*) FROM job_impression` takes _several minutes_ to return the current result (~43 million records) despite having a composite primary key that should have been used to return instantly. It's unclear why [Postgres is failing to use the index](https://wiki.postgresql.org/wiki/Index-only_scans).

2. For each impression, we update the viewcount in redis by running a sum operation. This could be slow.

3. Postgres has a typical CPU load of 50%, but it's unclear which queries are taxing. Query logs may help.

**Comment by jace** [comment]
Query cost for point 2, updating viewcount:

```
hasjob=> explain analyze select count(distinct(event_session.user_id)) from event_session, job_impression where event_session.user_id is not null and event_session.id = job_impression.event_session_id and job_impression.jobpost_id=34054;
                                                                            QUERY PLAN                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=42186.45..42186.46 rows=1 width=4) (actual time=1398.858..1398.859 rows=1 loops=1)
   ->  Hash Join  (cost=29640.43..42182.28 rows=1666 width=4) (actual time=1304.206..1392.682 rows=4638 loops=1)
         Hash Cond: (job_impression.event_session_id = event_session.id)
         ->  Index Only Scan using job_impression_pkey on job_impression  (cost=0.56..11473.91 rows=6247 width=4) (actual time=0.080..42.315 rows=5941 loops=1)
               Index Cond: (jobpost_id = 34054)
               Heap Fetches: 5941
         ->  Hash  (cost=25630.55..25630.55 rows=244345 width=8) (actual time=1303.881..1303.881 rows=246683 loops=1)
               Buckets: 16384  Batches: 4  Memory Usage: 2418kB
               ->  Bitmap Heap Scan on event_session  (cost=6058.10..25630.55 rows=244345 width=8) (actual time=57.853..1004.216 rows=246683 loops=1)
                     Recheck Cond: (user_id IS NOT NULL)
                     Heap Blocks: exact=17102
                     ->  Bitmap Index Scan on ix_event_session_user_id  (cost=0.00..5997.01 rows=244345 width=0) (actual time=53.699..53.699 rows=251925 loops=1)
                           Index Cond: (user_id IS NOT NULL)
 Planning time: 0.574 ms
 Execution time: 1635.642 ms
```

**Comment by jace** [comment]
The query plan suggests the hash join is used because the planner felt that was appropriate, but is clearly not -- 29.6 seconds estimated, only 1.3 seconds used. This post suggests [hash join was a bad choice by the query planner](http://serverfault.com/questions/502198/why-is-hash-join-slow-in-postgresql-9-1) and running `VACUUM ANALYZE` on the table may help update the query planner's statistics.

**Comment by jace** [comment]
In 3ab7fa1 we no longer hit the database to perform what is obviously a current_value+1 operation.

**Comment by jace** [comment]
In #233 the new indexed `datetime` column improves performance of issue 1 above (performance of `COUNT(*)` on the table).

Before:

```
explain analyze select count(*) from job_impression;

                                                              QUERY PLAN                                                               
---------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=1128869.66..1128869.67 rows=1 width=0) (actual time=308371.802..308371.803 rows=1 loops=1)
   ->  Seq Scan on job_impression  (cost=0.00..1003870.13 rows=49999813 width=0) (actual time=0.010..260229.584 rows=43604781 loops=1)
 Planning time: 0.077 ms
 Execution time: 308371.867 ms
```

After:

```
explain analyze select count(*) from job_impression;

                                                                                 QUERY PLAN                                                                                  
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=1523417.39..1523417.40 rows=1 width=0) (actual time=12110.181..12110.181 rows=1 loops=1)
   ->  Index Only Scan using ix_job_impression_datetime on job_impression  (cost=0.56..1426844.12 rows=38629304 width=0) (actual time=1.944..8319.523 rows=38631658 loops=1)
         Heap Fetches: 0
 Planning time: 0.412 ms
 Execution time: 12110.311 ms
```

**Comment by jace** [comment]
In PostgreSQL, `DISTINCT` returns distinct non-null values only, so the where clause `event_session.user_id is not null` is not necessary. Removing it improves performance dramatically:

```
explain analyze select count(distinct(event_session.user_id)) from event_session, job_impression where event_session.id = job_impression.event_session_id and job_impression.jobpost_id=33349;

                                                                        QUERY PLAN                                                                        
----------------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=31135.37..31135.38 rows=1 width=4) (actual time=0.067..0.067 rows=1 loops=1)
   ->  Nested Loop  (cost=0.99..31122.85 rows=5008 width=4) (actual time=0.027..0.054 rows=4 loops=1)
         ->  Index Only Scan using job_impression_pkey on job_impression  (cost=0.56..252.20 rows=5008 width=4) (actual time=0.015..0.019 rows=4 loops=1)
               Index Cond: (jobpost_id = 33349)
               Heap Fetches: 0
         ->  Index Scan using event_session_pkey on event_session  (cost=0.42..6.15 rows=1 width=8) (actual time=0.006..0.007 rows=1 loops=4)
               Index Cond: (id = job_impression.event_session_id)
 Planning time: 0.427 ms
 Execution time: 0.190 ms
(9 rows)
```

**Comment by jace** [comment]
Using the fix in ff53282 and reversing the hack in 3ab7fa1 (which is vulnerable to propagating counting errors), I believe we've resolved the performance problem. Production server load has dropped to 2.x this morning from the past month's average of 5-7.x

However, recounting impression stats after every impression is still inefficient as (a) we're not interested in an immediate high accuracy count and (b) multiple visitors in parallel will cause a lot of recounting. Instead, the `jobpost_id` should be added to a dirty set and the impressions should be recounted by a periodic job (see #221 for periodic job runner requirement) running every minute (but not overlapping itself) or running in a continuous loop. This requires a separate ticket.

**Comment by jace** [comment]
We've somehow lost the performance improvement from #233 [as described above](#issuecomment-152861440). `COUNT(*) FROM job_impression` is doing a sequential scan again. Vacuuming the table hasn't helped.

```sql
hasjob=> explain analyze select count(*) from job_impression;

                                                              QUERY PLAN                                                              
--------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=1321332.35..1321332.36 rows=1 width=0) (actual time=22003.833..22003.833 rows=1 loops=1)
   ->  Seq Scan on job_impression  (cost=0.00..1162772.48 rows=63423948 width=0) (actual time=8.023..14931.370 rows=63423947 loops=1)
 Planning time: 2.633 ms
 Execution time: 22003.977 ms
(4 rows)
```

**Comment by jace** [comment]
Various readings suggest PostgreSQL will revert to a sequential scan from an index scan if the index is too large. Since partitioning is the only solution here, I'm closing this issue to continue discussion in #309.

* [Why Postgres Won't Always Use an Index](https://robots.thoughtbot.com/why-postgres-wont-always-use-an-index)
* [Fast Counting](http://rhaas.blogspot.in/2011/10/fast-counting.html)
* [Why is my index not being used?](http://www.depesz.com/2010/09/09/why-is-my-index-not-being-used/)

**Comment by jace** [comment]
[Found the explanation](https://news.ycombinator.com/item?id=11877752). `COUNT(*)` uses the index only if the table has been vacuumed, which is not instant. Therefore `COUNT(*)` is inefficient at just the times we depend on it.
